### PR TITLE
Fix a set of memory leaks uncovered by address sanitizer

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Fixes issues reported by the leak checker in address sanitizer in the following libraries and tests:
* Geometry
* NexusGeometry
* DataObjects
* DataHandling

The aim is to get a clean run through address sanitizer and have a Jenkins job running this. This is part 1 of the set of fixes.

**To test:**

Code review. I would suggest reviewing each commit in turn.

If you would like to run the tests then you can use the `WITH_ASAN` cmake option but you will need to create a file `lsan.supp` with the following contents

```
# NeXus has a known memory leak when reading attributes.
# It is fixed upstream but not made it to distribution packages
# Direct leak of 12114 byte(s) in 1346 object(s) allocated from:
#     #0 0x7fb126db5538 in strdup (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x77538)
#     #1 0x7fb120a534cb  (/usr/lib/libNeXus.so.0+0xc4cb)
#     #2 0x7fb100000000  (<unknown module>)
leak:libNeXus

# OpenCascade memory allocation seems to confuse ASAN
# Direct leak of 544 byte(s) in 1 object(s) allocated from:
#     #0 0x7f77eec3a618 in operator new[](unsigned long) (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xe0618)
#     #1 0x7f77e4141b08  (/usr/lib/x86_64-linux-gnu/libTKernel.so.11+0x7bb08)
leak:libTKernel

# TBB leaks some memory allocated in singleton depending on deinitialization order
# Direct leak of 1560 byte(s) in 3 object(s) allocated from:
#     #0 0x7f7ae72a80a0 in operator new[](unsigned long) (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc80a0)
#     #1 0x7f7ae595d13e  (/usr/lib/x86_64-linux-gnu/libtbb.so.2+0x2213e)
leak:libtbb

# Python and associated libraries *appears* to leak memory.
leak:libpython2.7
leak:umath
leak:multiarray
leak:_ctypes
``` 

and set `export LSAN_OPTIONS="suppressions=\"lsan.supp\""` to point to this file before running the tests. This will be committed to the codebase when everything is fixed and the file content is fully known.

*There is no associated issue.*

*This does not require release notes* because **there is nothing concrete to report.**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/MDAlgorithms/test/ConvertToQ3DdETest.h [hunk 1]
```diff
@@ -407,16 +389,13 @@ class ConvertToQ3DdETest : public CxxTest::TestSuite {
   void xtestTransf_GL() {
```
Comment (DanNixon) [inline Framework/MDAlgorithms/test/ConvertToQ3DdETest.h]: Are all of these tests disabled (prefixed with `x`)? Do they not pass?

(I know this is not relevant to the work here, just worth mentioning)

File: Framework/DataHandling/inc/MantidDataHandling/XmlHandler.h [hunk 2]
```diff
@@ -4,43 +4,25 @@
...
-#include <Poco/DOM/NodeIterator.h>
-#include <Poco/DOM/NodeList.h>
-#include <Poco/DOM/Text.h>
-#include <Poco/SAX/InputSource.h>
-
 namespace Mantid {
 namespace DataHandling {
 
-class DLLExport XmlHandler {
+class MANTID_DATAHANDLING_DLL XmlHandler {
```
Comment (DanNixon) [inline Framework/DataHandling/inc/MantidDataHandling/XmlHandler.h]: :+1:

**Review by DanNixon — COMMENTED** [review]
Comments I had for the first set of commits (I started reviewing this before the CI failed).

File: Framework/Crystal/src/FindUBUsingLatticeParameters.cpp [hunk 3]
```diff
@@ -105,10 +105,10 @@ void FindUBUsingLatticeParameters::exec() {
   {      // and save the UB in the sample
     char logInfo[200];
     int num_indexed = IndexingUtils::NumberIndexed(UB, q_vectors, tolerance);
+    // cppcheck-suppress invalidPrintfArgType_uint
     sprintf(logInfo,
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/Crystal/src/FindUBUsingLatticeParameters.cpp]: :+1:

File: Framework/API/test/SampleTest.h [hunk 4]
```diff
@@ -76,26 +76,24 @@ class SampleTest : public CxxTest::TestSuite {
...
-    TS_ASSERT_THROWS_NOTHING(sample.setOrientedLattice(latt));
+    auto lattice = std::make_unique<OrientedLattice>(1.0, 2.0, 3.0, 90, 90, 90);
+    auto latticeAddress = lattice.get();
+    TS_ASSERT_THROWS_NOTHING(sample.setOrientedLattice(std::move(lattice)));
 
     const OrientedLattice &retLatt = sample.getOrientedLattice();
     // Test that this references the correct object
-    // TS_ASSERT_EQUALS(&retLatt, latt);//This is no longer correct.
-    // setOrientedLattice makes a copy of the OrientedLattice object
+    TS_ASSERT_EQUALS(&retLatt, latticeAddress); // This is no longer correct.
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/API/test/SampleTest.h]: Good catch. I'll update the comment to describe the test properly. It's not necessarily obvious at first glance.

File: Framework/MDAlgorithms/test/ConvertToQ3DdETest.h (hunk 1 above)
Comment (martyngigg) [reply Framework/MDAlgorithms/test/ConvertToQ3DdETest.h]: Good catch! I was so focused on the changes I hadn't spotted the `x`. I'd love to know why we have disabled tests here. I'll try and find out why.

**Comment by martyngigg** [comment]
@DanNixon I should have fixed all of the issues here now.

**Review by DanNixon — APPROVED** [review]
:+1:

**Comment by martyngigg** [comment]
Thanks for reviewing this quickly,

@mantidproject/gatekeepers Could we get this merged as there are a fair few changes across a lot of files.

**Pull request merged**